grafana/grafana
# Loki: Refactor template variable logic and remove reliance on `timeSrv`

Part of https://github.com/grafana/grafana/issues/72631 and following up on https://github.com/grafana/grafana/pull/78450.

In this pull request, I aim to eliminate duplicated logic related to querying for labels, values, and series. By leveraging methods implemented in languageProvider, we can benefit from its logic, which handles the use of `defaultTimeRange` when `timeRange` is not provided as a parameter. One of the goals is to keep  `datasource.languageProvider.getDefaultTimeRange()` `private` and avoid its usage in `datasource.ts` if unnecessary.

Here are the key changes:

**1. Use `languageProvider` methods to fetch labels, values, and series instead of duplicating logic in `datasource.ts`**

Previously, we had redundant methods like `labelNamesQuery`, l`abelValuesQuery`, and `labelValuesSeries` that essentially duplicated the functionality of `fetchLabels`, `fetchLabelValues`, and `fetchSeriesLabels` in the language provider. The overall objective is to eliminate the need for `timeSrv` and the potential duplication of logic regarding the default time range. Fortunately, we can achieve this by reusing methods in the language provider that handle cases when `timeRange` is not provided. This change not only simplifies the code but also ensures a single source of truth for fetching labels, values, and series across the Loki codebase.

**2. Unify querying logic in `metricFindQuery`**

In `metricFindQuery`, we previously had logic that directed variable queries through two different querying methods based on whether the variable query was a string or an object. In this pull request, I introduced a new method, `parseStringToVariableQuery`, which converts strings to query models. Subsequently, we use a unified logic for querying, streamlining the process.

**3. In `getTagKeys`, `getTagValues`, and `metricFindQuery`, use the passed `timeRange`**

All these methods receive an optional timeRange parameter. While it was previously unused, we now take advantage of this timeRange.

**4. Non-related, but done while working on the files**

Rename `variables.ts` to `LokiVariableSupport.ts` following the recommended file-naming convention for files with exported classes.

As evident, many of these changes are implementation details and do not alter the final results. Consequently, tests are passing, ensuring that label, value, and series queries are still correctly executed through `datasource.metadataRequest` in the end.

**How to test:**

I have created this dashboard with variables created in string format (in version 8.5) - e.g. `label_names()`, `label_values(label)`: https://gist.github.com/ivanahuckova/5fed6b862d62ca06316fe613d53df516. Import the dashboard and check that everything is correctly rendered.
